### PR TITLE
add new module: window_manipulation

### PR DESCRIPTION
**Pull request by lasers**
Hi, I made ~~`window_list`.~~ `window_manipulation`.

*- Color icons... based on scratchpad, floating, focused, urgent, etc*
2017-08-08-150549_3840x1200_scrot
*- Rainbow title on all windows*
2017-08-12-111946_793x69_scrot
*- Rainbow border pixel on all windows (That zero is Netflix)*
2017-08-12-130304_1100x109_scrot
*- Danger block on all windows... based on window percent*
2017-08-12-131031_1095x107_scrot

... and many more. 😱

### CODE REVIEW

File: py3status/modules/window_list.py [hunk 1]
```diff
@@ -0,0 +1,228 @@
...
+                new_scratchpad.append(s.id)
+
+            for w in windows:
+                is_scratchpad = False
+                if w.id in new_scratchpad:
+                    is_scratchpad = True
+
+                new_windows.append({
+                    'wid': w.id,
+                    'name': w.name,
```
Comment (tobes) [inline py3status/modules/window_list.py]: `'class': w.window_class,`? this seems useful

File: py3status/modules/window_list.py [hunk 2]
```diff
@@ -0,0 +1,228 @@
+# -*- coding: utf-8 -*-
+"""
+Display window list asynchronously.
+
```
Comment (lasers) [inline py3status/modules/window_list.py]: Fixed.

File: py3status/modules/window_list.py [hunk 3]
```diff
@@ -0,0 +1,228 @@
...
+Configuration parameters:
+    button_focus: mouse button to focus a window (default 1)
```
Comment (lasers) [inline py3status/modules/window_list.py]: Users can scroll windows with mouse, click to focus. (disabled by default).

File: py3status/modules/window_list.py [hunk 4]
```diff
@@ -0,0 +1,228 @@
...
+    {scratchpad}     number of scratchpad windows    eg 4
+    {urgent}         number of urgent windows        eg 2
+    {window}         number of windows               eg 11
+
+format_window placeholder:
+    {name}           window title name
+    is_floating      a boolean based on window data
+    is_focused       a boolean based on window data
+    is_scratchpad    a boolean based on window data
+    is_urgent        a boolean based on window data
```
Comment (lasers) [inline py3status/modules/window_list.py]: This is explicitly boolean (and not placeholders). See format's placeholders. They don't have `is_` because it can be used still... to show numbers. This is only good for controlling how things should appear... so I'd like to make them more clear... and is helpful in organizing (eg, they all stay together because of `is_`...). Same as `gcal` and some other modules.

File: py3status/modules/window_list.py [hunk 5]
```diff
@@ -0,0 +1,228 @@
...
+    """
+    """
+    # available configuration parameters
+    button_focus = 1
+    button_next = None
+    button_previous = None
+    format = '{format_window}'
+    format_separator = ' '
+    format_window = u'\?max_length=30 [\?if=is_urgent&color=urgent \u2756' +\
+        '|[\?if=is_focused&color=focused \u2756|\?color=unfocused \u2756]] {name}'
```
Comment (lasers) [inline py3status/modules/window_list.py]: This would break the full customization (eg different icons and/or names).

File: py3status/modules/window_list.py (hunk 1 above)
Comment (lasers) [inline py3status/modules/window_list.py]: Added. :-)

File: py3status/modules/window_list.py [hunk 6]
```diff
@@ -0,0 +1,228 @@
...
+                is_scratchpad = False
+                if w.id in new_scratchpad:
+                    is_scratchpad = True
+
+                new_windows.append({
+                    'wid': w.id,
+                    'name': w.name,
+                    'is_focused': w.focused,
+                    'is_urgent': w.urgent,
+                    'is_floating': w.floating,
```
Comment (lasers) [inline py3status/modules/window_list.py]: Fixed. I got lazy and stopped testing things under 2.  :-(

File: py3status/modules/window_list.py [hunk 7]
```diff
@@ -0,0 +1,228 @@
...
+                    w['is_floating'] = True
+                    floating += 1
+                else:
+                    w['is_floating'] = False
+            except:
+                pass
+            new_data.append(self.py3.safe_format(self.format_window, w))
+
+        self.window, self.floating, self.scratchpad, self.urgent = \
+            window, floating, scratchpad, urgent
```
Comment (lasers) [inline py3status/modules/window_list.py]: Cleaned up. 👍

File: py3status/modules/window_list.py [hunk 8]
```diff
@@ -0,0 +1,228 @@
...
+            'cached_until': self.py3.CACHE_FOREVER,
+            'full_text': self.py3.safe_format(self.format,
+                                              dict(
+                                                  floating=self.floating,
+                                                  format_window=self.fmt_window,
+                                                  scratchpad=self.scratchpad,
+                                                  urgent=self.urgent,
+                                                  window=self.window
+                                              ))
+        }
```
Comment (lasers) [inline py3status/modules/window_list.py]: Cleaned up. 👍

File: tests/test_module_doc.py [hunk 9]
```diff
@@ -35,6 +35,7 @@
 # Ignored items will not have their default values checked or be included for
 # alphabetical order purposes
 IGNORE_ITEM = [
+    ('window_list', 'format_window'),  # line too long for docstring parsing
```
Comment (lasers) [inline tests/test_module_doc.py]: This would break the full customization (eg different icons and/or names)... as well as controlling how windows should appear.

File: py3status/modules/window_list.py (hunk 5 above)
Comment (tobes) [inline py3status/modules/window_list.py]: ok but I think you are resigning this module to a long struggle to get merged.

this does not sound like a feature currently needed, so could be added later as improvement.  I'm also of the view that there is a fundamental broken logic going on here, because I can be urgent and focused etc.

**Comment by lasers** [comment]
>comments and docstrings are nice to have

Cleaned up. Docstrings where?

>screenshots are nice but good to show the default format too

Done. The default is plain white so you can see why I skipped that one.

>do we want a way to limit the windows shown?

Isn't that what the `format_window` is for... Put `{name}` inside `]]]` and such.

>I'm not sure what the buttons events do but maybe because it didn't work (fwiw I only have i3ipc installed for python2)

You configured `button_next`, `button_previous`? The default does not come with them enabled, but `show everything` option does... You can scroll windows all the way left or right with mouse, then click to focus. Works with urgent windows too.

Thank you for reviewing so fast. 👍 😆 :-)

**Comment by tobes** [comment]
> Docstrings where?

It is nice to know what functions are doing.  if nothing else I can then review them.  
eg `_manipulate_and_push_data(self, windows)` what is this doing and why?


> The default is plain white so you can see why I skipped that one.

the default is the most interesting.  Actually I was meaning in the PR

>  > do we want a way to limit the windows shown?
>
> Isn't that what the format_window is for... Put {name} or inside ]]] and such.

I meant like only showing some windows,  but anyhow that could come later if people desired

buttons:

I think these are a bit unclear.  when I clicked on a window name I expected it to focus from the description.  I think the buttons need a bit more thought.

**Comment by lasers** [comment]
>It is nice to know what functions are doing. if nothing else I can then review them.
eg _manipulate_and_push_data(self, windows) what is this doing and why?

Documented. :-)

>>>do we want a way to limit the windows shown?

>>Isn't that what the format_window is for... Put {name} or inside ]]] and such.

>I meant like only showing some windows, but anyhow that could come later if people desired

I included it in an example. See `# show all class names excluding floating and scratchpad` 😱 

>I think these are a bit unclear. when I clicked on a window name I expected it to focus from the description. I think the buttons need a bit more thought.

Yes! That's what it should be doing... Can you try clicking on windows on different workspaces since I think you're on laptop and may be clicking on window names that's right in front of you? 
  

>>The default is plain white so you can see why I skipped that one.

>the default is the most interesting. Actually I was meaning in the PR

As you can see, I renamed it to `window_manipulation` as I don't want to dwell on the name `list`. It's not a `list`. It can behave like `window_list_async`, `window_title_async`, or `scratchpad_async`... and possibly more.

I think I might want to make `format`, `format_window`, and `button_click` empty... basically to start everything with nothing... and users can start from copying examples... and find what they like. Thoughts?

Reason:  The default `format_window` shows everything.. and you wanted some windows excluded... I don't want to default anything, really. :-)

I cleaned up the module some more... Organized things more... in a way that it makes sense. I hope things will be more clear now. Cheers. 👍

File: py3status/modules/window_manipulation.py [hunk 10]
```diff
@@ -0,0 +1,396 @@
...
+    """
+    """
+    # available configuration parameters
+    allow_urgent = None
+    button_focus = None
+    button_next = None
+    button_previous = None
+    format = '{format_window}'
+    format_separator = ' '
+    format_window = '[\?if=is_focused&color=good \u2022|\?color=bad \u2022]'
```
Comment (tobes) [inline py3status/modules/window_manipulation.py]: `u'` we care about python 2

The info is quite limited here I like `u'[\?if=is_focused&color=good \u2022|\?color=bad \u2022] {class}:[\?max_length=20 {name}]'`

File: py3status/modules/window_manipulation.py [hunk 11]
```diff
@@ -0,0 +1,396 @@
...
+from threading import Thread
+
+import i3ipc
+
+
+class Py3status:
+    """
+    """
+    # available configuration parameters
+    allow_urgent = None
```
Comment (tobes) [inline py3status/modules/window_manipulation.py]: `allow_urgent` is a restricted name and is actually used universally (as every documentation could be improved).  Also the description was not clear to the functionality

File: py3status/modules/window_manipulation.py [hunk 12]
```diff
@@ -0,0 +1,396 @@
...
+        t.daemon = True
+        t.start()
+
+    def _loop(self):
+        def update_window_manipulation(i3, e=None):
+            scratchpads = i3.get_tree().scratchpad().leaves()
+            windows = i3.get_tree().leaves()
+
+            # for w in windows:
+            #     print('OUTPUT... %s' % w.border)
```
Comment (tobes) [inline py3status/modules/window_manipulation.py]: this looks like it is here by error

File: py3status/modules/window_manipulation.py [hunk 13]
```diff
@@ -0,0 +1,396 @@
...
+        i3.on('window::close', update_window_manipulation)
+        i3.on('window::focus', update_window_manipulation)
+        i3.on('window::title', update_window_manipulation)
+        i3.on('window::fullscreen_mode', update_window_manipulation)
+        i3.on('window::move', update_window_manipulation)
+        i3.on('window::floating', update_window_manipulation)
+        i3.on('window::urgent', update_window_manipulation)
+        i3.on('window::mark', update_window_manipulation)
+        i3.on('workspace::focus', update_window_manipulation)
+        i3.main()
```
Comment (tobes) [inline py3status/modules/window_manipulation.py]: with this function i'd do
```
def _loop(self):
    try:
        ...
    except:
        self.py3.report_exception('monitor thread error')
```

otherwise we do not see the errors if they happen in the thread which is a pain to debug

File: py3status/modules/window_manipulation.py [hunk 14]
```diff
@@ -0,0 +1,396 @@
...
+        format_separator = self.py3.safe_format(self.format_separator)
+        self.fmt_window = self.py3.composite_join(format_separator, new_data)
+        self.py3.update()
+
+    def _switch_window_manipulation(self, direction):
+        """
+        Users can scroll windows with mouse, click to focus. If an urgent occurs,
+        we will use a different click to focus. Disabled by default.
+        """
+        windows, f = self.shared_data, 'is_focused'
```
Comment (tobes) [inline py3status/modules/window_manipulation.py]: `f` makes this function hard to read just use the literal.

also this is an evil double assignment thing which are just bad like this 

but you could consider simplifying things in this function eg
```diff
diff --git a/py3status/modules/window_manipulation.py b/py3status/modules/window_manipulation.py
index 0f4477a..0672952 100644
--- a/py3status/modules/window_manipulation.py
+++ b/py3status/modules/window_manipulation.py
@@ -203,6 +203,7 @@ class Py3status:
         t = Thread(target=self._loop)
         t.daemon = True
         t.start()
+        self.active = 0
 
     def _loop(self):
         def update_window_manipulation(i3, e=None):
@@ -287,10 +288,10 @@ class Py3status:
 
         new_data = []
         for w in new_window:
-            count_window += 1
             # send all values (for focused_window) out to the format (as placeholders)
             # so the users can do some simple things like printing name only.
             if w['is_focused']:
+                self.active = count_window
                 self.id = w.get('id')
                 self.window_name = w.get('name')
                 self.class_name = w.get('class')
@@ -299,6 +300,8 @@ class Py3status:
             if w['is_urgent']:
                 count_urgent += 1
 
+            count_window += 1
+
             # when scrolling, we want to ignore this error because we already
             # fixed it on update. we turned this into a boolean too.
             try:
@@ -330,22 +333,10 @@ class Py3status:
         Users can scroll windows with mouse, click to focus. If an urgent occurs,
         we will use a different click to focus. Disabled by default.
         """
-        windows, f = self.shared_data, 'is_focused'
-        length = len(windows)
-        for index, w in enumerate(windows):
-            if w.get(f):
-                windows[index][f] = False
-                if direction < 0:  # switch previous
-                    if index > 0:
-                        windows[index - 1][f] = True
-                    else:
-                        windows[index][f] = True
-                elif direction > 0:  # switch next
-                    if index < (length - 1):
-                        windows[index + 1][f] = True
-                    else:
-                        windows[length - 1][f] = True
-                break
+        windows = self.shared_data
+        self.active = (self.active + direction) % len(windows)
+        for index, window in enumerate(self.shared_data):
+            window['is_focused'] = (index == self.active)
 
         self._manipulate_and_refresh(new_window=windows)
 
```
slightly changes behaviour as we now loop round

File: py3status/modules/window_manipulation.py [hunk 15]
```diff
@@ -0,0 +1,396 @@
...
+
+        # put everything together and do a refresh
+        format_separator = self.py3.safe_format(self.format_separator)
+        self.fmt_window = self.py3.composite_join(format_separator, new_data)
+        self.py3.update()
+
+    def _switch_window_manipulation(self, direction):
+        """
+        Users can scroll windows with mouse, click to focus. If an urgent occurs,
+        we will use a different click to focus. Disabled by default.
```
Comment (tobes) [inline py3status/modules/window_manipulation.py]: >  If an urgent occurs, we will use a different click to focus.

I don't understand this

also this just seems to change the 'active' window so a nicer method name would make sense

File: py3status/modules/window_manipulation.py [hunk 16]
```diff
@@ -0,0 +1,396 @@
...
+
+    def _loop(self):
+        def update_window_manipulation(i3, e=None):
+            scratchpads = i3.get_tree().scratchpad().leaves()
+            windows = i3.get_tree().leaves()
+
+            # for w in windows:
+            #     print('OUTPUT... %s' % w.border)
+
+            new_scratchpad, new_window = self._organize_data(scratchpads, windows)
```
Comment (tobes) [inline py3status/modules/window_manipulation.py]: `_organize_data` should probably just be inline here now that you have re-factored.

**Comment by tobes** [comment]
Thanks,

much better with comments etc even if slightly snarky :)

I prefer the old name.  because it is info about windows.

ok I was expecting if I clicked on a window name in the i3bar it would get the focus eg like a taskbar would

File: py3status/modules/window_manipulation.py (hunk 11 above)
Comment (lasers) [inline py3status/modules/window_manipulation.py]: > Also the description was not clear to the functionality

Take a guess. :-) Okay, I'll update the description.

* `window` allows all urgents including ordinary `scratchpad` workspace.
* `scratchpad` allows urgent in scratchpad only. This is so we can behave like scratchpad properly without getting an urgent for something that happens in `window` (eg any workspace).

Maybe... `all`, `scratchpad` would be better.

What if I start off checking for a boolean instance... Otherwise, strings... Would that work okay? It's off by default (ie None) because it's ugly and unproductive to support `urgent` by default... is which why I included a fake `urgent` example.

File: py3status/modules/window_manipulation.py (hunk 10 above)
Comment (lasers) [inline py3status/modules/window_manipulation.py]: I'll include this in a list of examples... as `keep it short and simple`.

I don't want to dwell on the name `list`. It's not a `list`. It can behave like `window_list_async`, `window_title_async`, or `scratchpad_async`... and possibly more.

This makes `list` a default format. I'd like `None` for `format` and `format_window`, but I don't know if that's a good idea, so I use dots.... Thoughts on dots and/or empty variables? Users should have to put in something they want... instead of using default.

Comment (tobes) [inline py3status/modules/window_manipulation.py]: but the default format is just a list of dots which doesn't seem useful for most users

File: py3status/modules/window_manipulation.py (hunk 15 above)
Comment (lasers) [inline py3status/modules/window_manipulation.py]: I `s/different/urgent` in this description. Urgent is internally primary.
```
If urgent:
---- click to focus (urgent)
else:
---- click to focus (selected window id).

File: py3status/modules/window_manipulation.py (hunk 16 above)
Comment (lasers) [inline py3status/modules/window_manipulation.py]: Can you explain more? Do you mean this?
```
            new_scratchpad, new_window = self._organize_data(i3.get_tree().scratchpad().leaves(), windows = i3.get_tree().leaves())
```

I might want to keep it as-is... so we can internally test the new values and see if they're usable before adding... (See the accidental commented out code above)

Comment (tobes) [inline py3status/modules/window_manipulation.py]: I mean just have all the code that is in `self._organize_data()` here don't put it in it's own function.

It is really only worth having functions if

a) the code is called from multiple places

b) you are encapsulating complicated stuff

c) making things more readable or easier to debug.

`update_window_manipulation()` is the only place we call `_organize_data()`, and that's all we do apart from then calling `_manipulate_and_refresh()` at the end

File: py3status/modules/window_manipulation.py (hunk 14 above)
Comment (lasers) [inline py3status/modules/window_manipulation.py]: No `windows` for `self._manipulate_and_refresh(new=window=windows)`.

This is for scrolling only...

When we start scrolling, we get the real data first time.... then we start faking things... and everytime we scroll (still on the bar), we continue to work with fake data in the loop (because of no new events... so no new data)

If events changes, we would just work with that data... then fake again. This function does not know which one is real or not... since we just shift things... then push out via the same `self._manipulate_and_refresh(new_window=windows)`.

Short version... Things are fake as shit when we scroll.

File: py3status/modules/window_manipulation.py (hunk 16 above)
Comment (lasers) [inline py3status/modules/window_manipulation.py]: `c`. making things more readable or easier to debug. I moved all that into `organize`. 

We won't have to look at...
* `update_window_manipulation`   # get the damn windows and scratchpad
* `window_manipulation` # print the damn thing. urgent too if needed.

We only have to look...
* `_organize_data` # figure out what we need, fix them up, get them all in one place.
* `_manipulate_data` # make counters, fix names/values, flip shit, make new data happen.

File: py3status/modules/window_manipulation.py (hunk 14 above)
Comment (tobes) [inline py3status/modules/window_manipulation.py]: `self._manipulate_and_refresh(new_window=self.shared_data)` should fix it.

Anyway currently things are much more confused than they need to be, simplifying stuff is fun :)

Also please try to spend a little more time on these comments too.  I'm not sure what you are saying here eg `If we click something` has no meaning.

you also say if we scroll we have no data but we have new data?

Comment (lasers) [inline py3status/modules/window_manipulation.py]: When we scroll, we work with the last updated data.... The only time we get new data is when those events occurs... usually involving us leaving the bar... and something get updated.

There could be some cases where we get updates during scrolling (like title change and urgent change and... whatnot, but this kind of issue is super trivial).

When we scroll, we send the data back to ourselves (or via shared_data).

If you don't understand urgent part, do `sleep 3 && echo -e "\a"`, then send it somewhere else... different workspace... or scratchpad.  Also, play with `allow_urgent`... so you can see how urgent in current workspace does not trigger the `window_manipulation` unless it's a scratchpad window.

Comment (tobes) [inline py3status/modules/window_manipulation.py]: please note `allow_urgent` is a restricted parameter so it __must not__ be used in your module.

for any module if `allow_urgent` is False then any urgency is removed.

So you need to at least rename it :)

also i3bar events (which are the real events from py3status point of view) are different from i3 events so try to be clear which you refer to on this PR as we have both types.

anyhow I still think this module is fun but could be made simpler / more readable

Comment (lasers) [inline py3status/modules/window_manipulation.py]: Quick Q. Can we be clever here in the module... and still retain the name?
```
if instance (boolean) self.allow_urgent
      if is False
          urgent = False
elif instance (string) self.allow_urgent
.... etc.
```
Any suggestions on the name? I got `urgent_type`...

Comment (tobes) [inline py3status/modules/window_manipulation.py]: Quick answer no.

To be honest I'm still not sure what your doing here with the urgent stuff or more importantly why.  I haven't actually looked in detail as the code has other issues imho

**Comment by lasers** [comment]
@tobes I pretty much finished adding the missing stuffs for the scratchpad part. There probably are some rough edges in there that you could look through and simplify if possible.... I also added lot of new examples... for scratchpad. 😄 

Two bugs worth mentioning... That `i3` and/or `i3ipc` freeze bug would occur... fairly rarely, but occasionally... and (some) data are just missing in `py3status2` and/or `i3-ipc`... which is why I think things might not make sense for you... if you're using python2.

New bursts of reviews, please! 👍 :-)

**Comment by tobes** [comment]
> Two bugs worth mentioning... That i3 and/or i3ipc freeze bug can occur... It's fairly rarely, but still occasional... and (some) data are just missing in py3status2 and/or i3-ipc (python2)... which is why I think things might not make sense for you... if you're using that version.

could you please explain

File: py3status/modules/window_manipulation.py [hunk 17]
```diff
@@ -0,0 +1,477 @@
...
+    def post_config_hook(self):
+        if self.format_window is None:
+            self.format_window = ''
+        if self.format_scratchpad is None:
+            self.format_scratchpad = ''
+
+        self.fmt_scratchpad = self.fmt_window = self.id = \
+            self.window_name = self.class_name = self.which_data = None
+        self.count_floating = self.count_scratchpad = \
+            self.count_urgent_sp = self.count_urgent = self.count_window = 0
```
Comment (tobes) [inline py3status/modules/window_manipulation.py]: can you not do this I am stopping my review here because this is not readable.

please do not do this multiple assignment like this because it is horrible.  Also try to avoid using `\` to extend lines as it usually means the code is badly written

Comment (lasers) [inline py3status/modules/window_manipulation.py]: Updated without multiple assignments + `\` extending. 

1) The `Seemingly random i3 freezes (SIGSTOP) #2539` issue mentioned in i3
2) I think just `is_floating` issue. Other than that, it's 99% workable.
* Python2 fix `'is_floating': getattr(w, 'floating', False)`... returns `False` everywhere.
* Python3 returns correctly... either `auto_off`, `auto_on`, `user_on`, or `user_off`.

**Comment by lasers** [comment]
Hi. I finished adding new format + control placeholders. I pretty much added most of everything that we can use... including `{pixel}` and `{percent}`. :-)

Also, we now let users decide what events to subscribe... This is really really nice because the users now can adjust the update frequency easily, hence the `window_manipulation`. ;-)

I think I'm finished with everything. This probably could go in `3.6` after reviews and stuffs. Cheers. 👍

**Comment by tobes** [comment]
ok,  as this stands I'm not going to review it, sorry

It is far too complicated and I do not agree with the `is_xxx` in the formats especially.

comments like `# ----------` have no place in the code a blank line does as well.

The way to get a module like this into py3status imho is to start simply (like you did) and then slowly expand the functionality after it is merged.  Now you have added too much contentions stuff I do not see it moving anywhere fast.

The name should also have some meaning like `windows_info` that is more applicable.

Also this module currently feels user hostile to me.

**Comment by lasers** [comment]
>It is far too complicated and I do not agree with the is_xxx in the formats especially.

`is_` prefix is nice and short... and is for boolean only. They stay together.

>comments like # ---------- have no place in the code a blank line does as well.

Replaced w/ word comments. They stay together too, but separated (ie `border`, `layout`, etc).

>... and then slowly expand the functionality after it is merged. Now you have added too much contentions stuff I do not see it moving anywhere fast.

I don't expect the users to slowly expand this functionality. It's better this way that I look through everything now and... and add all usable placeholders (if I can)... Also, it's good that I think I already added pretty much everything usable.... because now... maybe there won't be anything new to add.

>Also this module currently feels user hostile to me.

Yeah. They need to go through the "fire and furylike the world has never seen" in `py3status` formatting stuffs like I did... And we still need documentations for this kind of stuffs too. 👍 

>The name should also have some meaning like windows_info that is more applicable.

This works with scratchpad too. Since this is a complicated module that can do so many things... I'd rather keep `window_manipulation` and have no default settings... so we don't play any favors here... and would be easier to maintain when we don't have default settings in place (to skew the data and/or bugs).

---

I have two issues so far that I'd like to talk about.

1) Try `# show window list` manipulation
```
    format = '{format_window}'
    format_window = '\?max_length=30 '
    format_window += '[\?if=is_urgent [\?color=urgent&show \u2756] {name}'
    format_window += '|[[\?if=is_focused [\?color=focused&show \u2756] {name}'
    format_window += '|[\?if=is_scratchpad'
    format_window += '|[\?if=is_floating'
    format_window += '|]]]]]'

    events = ['window::focus', 'window::urgent']
```
You'll see a single window title (and you're able to scroll through window list). There are bunch of single whitespaces (from `|]]]]]`). I think that's a formatting bug we should fix.

2) My comment to you... in `L540` about color thresholds. I want the `index`, `pixel`, and `percent` in `format` just same as `format_window`... Different `format` contain different color values (focused color vs window loop color).

**Comment by tobes** [comment]
> is_ prefix is nice and short... and is for boolean only. 

I think this aspect needs more thought.  That's why I don't want to look at merging this stuff untill we work that out.  To be honest it works now and if you remove all traces of it from the module documentation it will work for you and just be a random hidden feature.  But maybe it will stop working because we make a change to the formatter.  Anyhow for me currently it is a total blocker.

> I don't expect the users to slowly expand this functionality.

me neither I'd expect __you__ to slowly make your monster without anyone noticing over a few PRs

> I have two issues..

your format looks funny, why not  `[\?if=a A]|[\?if=b B]|[\?if=c C]`

> L540 

not sure is this what you are after?

```
whatever = ???
safe_format(x, {'nice_name': whatever})
```

**Comment by lasers** [comment]
>me neither I'd expect you to slowly make your monster without anyone noticing over a few PRs

Yah. I thought I was done w/ making `window_list`, but then I saw that it could be more powerful and that it could do so much more... so I kept going.. then we now have this awesome ~~module~~ monster that can do so many things. 👍 

1) Blah! `format_separator = ''` fixes the issue... There were no formatting bug. *[sighs]*

2) I still have this issue. What if I already have a `{nice_name}` value? The issue could be that the `self.py3.threshold_get_color` wasn't designed to be used in chosen `formats`... like `self.py3.threshold_get_color(value, name, this_format_only_please)`.

    If you looked few lines down, you'll see this...
    ```
    # thresholds for window
    self.py3.threshold_get_color(w.get('percent'), 'percent')
    ```
    ... is where it get overwritten... and I think I did what you wanted, but with `\?color=%s %s` so the color is assigned immediately... but can't figure out how to make `format = '[\?color=percent {index} {name}` to work... so maybe it's a `self.py3.threshold_get_color` issue.

**Comment by lasers** [comment]
Progress report so far. Still working on this. Still nice. I kept run into occasional freezing i3 bugs ... probably from `i3ipc`. Also, I think to improve this some more... We should have something that can fix the (joining) `format_separator` issue... where it kept returning spaces.
* `'Window 1', '', '', '', 'Window 5'`... is to be converted into...
* `'Window 1', 'Window 5'`.

**Comment by ultrabug** [comment]
Friendly bump here?

**Comment by tobes** [comment]
Personally I think we should close this and then write a `i3wm` clean module that can do nice things and add features as we actually need them.

This module tries to do too much and is not very cleanly written imho.  We should start small.  Adding stuff because we can when there are no real users is crazy, and will just cause bitrot.
